Scheduler: OTel Instrumentation feature improvements and refactoring

mskacelik opened the pull request:
/cc @brunobat 
fixes: #28552
draft PR

I have Added OTel Instrumentation for quartz scheduling.
Currently there are two types of scheduling in quartz:
### Using the annotation `@Schedule`
For this type of scheduling there seems to be already existing feature for creating spans. Yet I feel that my implementation might be better because the spans contains information about the name of the executed method + duration of the method.
     
![image](https://github.com/quarkusio/quarkus/assets/124160830/4cdb11dc-9b9e-4cba-9a55-14a81ac49580)
> [NOTE] the first span (the top one) is my implementation which contains name of the method which was executed + duration `Thread.sleep(1000l);`
The second span (the bottom one) is the current implementation which will be activated when using property `quarkus.scheduler.tracing.enabled=true`

For some reason with my implemetation I cannot log `traceId` and `spanId` inside the executed method even tho span has been created (gonna have a look at it in a meantime).

### Using Quartz API (Programmatic) 
I have created feature which enables OTel instrumentation for each Job.
Currently the span creating is enabled by default if it finds the OTel extension in the project, but maybe we could use the property `quarkus.scheduler.tracing.enabled=true` for explicit enabling (like the current `@Schedule` approach).
Logging `traceId` + `spanId` works.

Basis for the code: https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/quartz-2.0/library/src/main/java/io/opentelemetry/instrumentation/quartz/v2_0

>[NOTE]: In the issue's reproducer there were few errors for the use of programmatic style scheduling like: 
>* not using injected scheduler (using the factory)
> * not having property for forced start

### Tests
Currently missing.

mskacelik committed:
Added OTel Instrumentation feature for both SimpleScheduler and Quartz Scheduler

Scheduler: introduce SPI module and JobInstrumenter

- implement JobInstrumenter in OTel extension
- remove the annotation transformer that adds OTel @WithSpan to
@Scheduled methods

added tests for scheduler/quartz OTel integration.

Co-authored-by: Martin Kouba <mkouba@redhat.com>

mskacelik commented:
Update:

Ok by double checking it seems that my implementation is the faulty one using the scheduler with `@Schedule` approach and the adding  `@WithSpan` via `quarkus.scheduler.tracing.enabled=true` which we have currently works as intended. But when it comes to programmatic scheduler my implementation works.

mkouba commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/tracing/WithSpanJob.java:
```diff
@@ -0,0 +1,53 @@
+    public WithSpanJob(Job delegate, boolean captureExperimentalSpanAttributes) {
+        this.delegate = delegate;
+        InstrumenterBuilder<JobExecutionContext, Void> jobInstrumenterBuilder = Instrumenter.builder(
+                CDI.current().select(OpenTelemetry.class).get(), "io.quarkus.opentelemetry",
```
It might make more sense to add an `OpenTelemetry` parameter to the constructor and provide the instance in the component that instantiates `WithSpanJob`...

mkouba commented on extensions/scheduler/common/src/main/java/io/quarkus/scheduler/common/runtime/WithSpanInvoker.java:
```diff
@@ -0,0 +1,50 @@
+    public WithSpanInvoker(ScheduledInvoker delegate) {
+        super(delegate);
+        InstrumenterBuilder<ScheduledExecution, Void> instrumenterBuilder = Instrumenter.builder(
+                CDI.current().select(OpenTelemetry.class).get(), "io.quarkus.opentelemetry",
```
Again, try to avoid `CDI.current().select()` if possible..

mkouba commented on extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java:
```diff
@@ -695,6 +706,13 @@ public boolean isBlocking() {
+        }
+
+        private boolean isOTelExtensionAvailable() {
+            return CDI.current().select(io.opentelemetry.api.OpenTelemetry.class).isResolvable();
```
I would try to avoid using `CDI.current().select(OpenTelemetry.class)` at all. In this particular case, I would just add the `Instance<OpenTelemetry>` injection point in the `SimpleScheduler` and use it here.

mkouba commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzSchedulerImpl.java:
```diff
@@ -757,6 +766,14 @@ private Optional<TriggerBuilder<?>> createTrigger(String identity, Scheduled sch
+    }
+
+    private static boolean isOTelExtensionAvailable() {
+        return CDI.current().select(io.opentelemetry.api.OpenTelemetry.class).isResolvable();
```
I would try to avoid using `CDI.current().select(OpenTelemetry.class)` at all. In this particular case, I would just add the `Instance<OpenTelemetry>` injection point in the SimpleScheduler and use it here.

radcortez reviewed (changes requested):
I'm concerned about giving visibility to OTel APIs into other extensions. Until now, we have tried to concentrate every OTel integration in the OTel extension. Because it is an optional dependency, these hacky workarounds are required to check if the OTel bean is available. 

I think it would be better to introduce a SPI Build Item that allows other extensions to customize the current extension (if Quarts or Schedule do not have another way already).  

For instance, we use a RESTEasy Reactive build item to register additional filters and do the OTel work, so RESTEasy doesn't have to know about OTel.

//cc @brunobat

brunobat commented:
> I'm concerned about giving visibility to OTel APIs into other extensions. Until now, we have tried to concentrate every OTel integration in the OTel extension. Because it is an optional dependency, these hacky workarounds are required to check if the OTel bean is available.
> 
> I think it would be better to introduce a SPI Build Item that allows other extensions to customize the current extension (if Quarts or Schedule do not have another way already).
> 
> For instance, we use a RESTEasy Reactive build item to register additional filters and do the OTel work, so RESTEasy doesn't have to know about OTel.
> 
> //cc @brunobat

`WithSpanJob implements Job` and inside it uses the OTel instrumentation API. 
There seems to be no scheduler side SPI to provide tracing, therefore it would be handy if we have one of our own... The Observation work also relates to this.
While we don't decide on it, I wonder if we could provide an SPI of our own to allow to isolate our extensions from this.

mskacelik commented:
My initial implementation was entirely within the **OTel** module. I was utilizing the `JobListener` interface, which managed the spans. Unfortunately, this approach did not work with Quarkus's `@Scheduler` and `JobDefinition` approach for creating scheduling tasks. To the best of my knowledge, the idiomatic approach for **Quarkus** involves using `SuccessfulExecution` and `FailedExecution` CDI events. However, with these events, it's impossible to distinguish between running jobs (missing JobExecutionContext).

mkouba commented:
> To the best of my knowledge, the idiomatic approach for **Quarkus** involves using `SuccessfulExecution` and `FailedExecution` CDI events. However, with these events, it's impossible to distinguish between running jobs (missing JobExecutionContext).

`SuccessfulExecution` and `FailedExecution` would not be enough. 

1. If I understand it correctly we need to execute some logic before the scheduled task is executed (start) and after the scheduled task is executed (end).
2. These events are not fired for Quartz jobs registered manually.

So the current behavior:
- If `quarkus.scheduler.tracing.enabled=true` and the OTel extension is present then the `@WithSpan` annotation is added automatically to every `@Scheduled` method
- Tasks registered programmatically with `JobDefinition` and Quartz jobs registered manually are not "traced automatically"

We could provide an SPI but we'll need some abstraction too because the Quartz jobs registered manually are not executed within the context of our API (i.e. no `io.quarkus.scheduler.ScheduledExecution`). In other words, we would need a `JobListener` that would bridge the SPI.

brunobat commented:
I've discussed this with @radcortez and we think a good solution would be to create 2 different SPI, one for the `WithSpanJob` and another one for the `WithSpanInvoker`.
This could be achieved either with a service loader or with dedicated `BuildItems`. 
A default impl for the interfaces would need to be present if OTel is not loaded. 
If Otel is present, it would create these beans... The place where this wrapping happens would need to be Beans. 
The OTel extension would have dependencies on these extensions.

brunobat commented:
Thanks @mkouba. It looks good to me. @mskacelik do you think you can add an integration test project for this PR?

radcortez commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzSchedulerImpl.java:
```diff
@@ -1039,9 +1064,16 @@ public Job newJob(TriggerFiredBundle bundle, org.quartz.Scheduler Scheduler) thr
+            return jobWithSpanWrapper(super.newJob(bundle, Scheduler));
+        }
+
+        private Job jobWithSpanWrapper(Job job) {
```
Maybe rename the method to not use concept names from OTel?

brunobat reviewed (dismissed):
We need and IT test project.

mkouba commented:
@mskacelik Any progress here? Do you need some help with the tests and docs?

mskacelik commented:
> @mskacelik Any progress here? Would you like some help with the tests and docs?

Hi, I started by creating a new project called `opentelemetry-quartz` in the `integration-tests` (I have not yet started working with the `SimpleScheduler` implementation).  I made three rest endpoints which initiate the three types of scheduling.
- `@Scheduled`
- JobDefinition
- custom job

> For each task, I also added a fixed `Thread.sleep(...)` to measure if the span duration matches the intended (at least or equal) `Thread.sleep` in the job.

The `quartz` project inspired these endpoints from the `integration-tests.` I also added some API from the `OpenTelemetry-spi` project, for example, exporting spans, resetting spans, and working with `InMemorySpanExporter`.

But for some reason, when I enable otel for quartz by: `quarkus.scheduler.tracing.enabled=true` on each created span, it throws some weird error. Note that I tried the current implementation in a separate Quarkus project, and it works as intended, so that might be some problem with the configuration of my `opentelemetry-quartz` module. Maybe you can help me with this @mkouba? I will be available at 12:00 to discuss it in zulip chat.

mskacelik commented:
# Tests
So, I added integration tests for Quartz. The only thing missing is Simple Scheduler tests, which will be pretty much the same as Quartz's (minus the custom job scheduler).

## In the test, there are 6 different schedulers:
1. **@Scheduled Type**
   - Normal scheduled task execution.
   - Scheduled task that intentionally throws an exception for testing.

2. **JobDefinition Type**
   - JobDefinition task execution.
   - JobDefinition task that intentionally throws an exception for testing.

3. **Custom Job Type (Programmatic)**
   - Custom job execution.
   - Custom job that intentionally throws an exception for testing.
   
Tests in the native mode do not work for some reason. Especially the `assertErrorJobSpan` where it checks on the error message.  It seems that spans that handle an exception do not contain the `message` of the exception as a field, nor do they contain stack trace. I tried to reproduce this problem in a separate Quarkus project, but everything worked fine, and spans in native mode contain the error message. I have no idea why this effect is only happening in my test.

mkouba commented on integration-tests/opentelemetry-quartz/src/main/resources/application.properties:
```diff
@@ -0,0 +1,11 @@
+quarkus.application.name=opentelemetry-quartz-instrumentation-it
+quarkus.application.version=999-SNAPSHOT
+
+disabled=disabled
```
What are these properties for?

mskacelik replied:
I borrowed these top properties from other OTel integration tests, but it should not affect the test's outcome.

mkouba commented on integration-tests/opentelemetry-quartz/pom.xml:
```diff
@@ -0,0 +1,144 @@
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy</artifactId>
```
I think that we should rather use `quarkus-resteasy-reactive` instead...

mkouba commented:
Merging as the test failures do not seem to be related (and fail in other PRs as well).

michalvavrik commented:
@mkouba @brunobat this changed OTEL operation names (see https://github.com/quarkus-qe/quarkus-test-suite/pull/1559) e.g. `SchedulerService.increment` => `1_io.quarkus.ts.opentelemetry.reactive.SchedulerService#increment`. That's a breaking change (if intentional? It doesn't look better) and as such should be in migration guide.

mkouba commented:
> 1_io.quarkus.ts.opentelemetry.reactive.SchedulerService#increment

`1_io.quarkus.ts.opentelemetry.reactive.SchedulerService#increment` is a generated job identity. The identity must be _unique_ and is either generated (this case) or specified with `@Scheduled.identity()`. `SchedulerService.increment` is not unique. So maybe we should use the FQCN + method suffix as the span name instead?

mskacelik commented:
@mkouba @michalvavrik I would personally have the job identifier as a span name because, for example, using `JobDefinition,` there is no `FQCN + method suffix` since the job has no class. That being said, I would add an explicit identity for the QE test because the default-generated identity does not have to be identical. So, there could be some nondeterministic behaviors.

mskacelik commented:
@mkouba @michalvavrik 
> If there is such a change and you can't rely on operation name, I think it needs to be documented.

Martin will write some documentation in the migration guide, something along the lines of:
> When tracing is enabled using *OpenTelemetry*, previously, the span names followed the `<classname>.<methodName>` format. This has been updated to use the *job's unique identifier* as its name.

michalvavrik commented:
> Well, the point is that the behavior was undefined/undocumented. So it wasn't testing a contract but an implementation detail.

I had to rely on something when writing these tests. I don't say you established contract, but little note in migration guide for people like me would help. Thanks

>  I mean, previously quarkus.scheduler.tracing.enabled=true implied that @WithSpan should be added to a scheduled method and the javadoc of @WithSpan.value() states: "If not specified, an appropriate default name should be created by auto-instrumentation. E.g. "className"."method"." Which is IMO not clear or deterministic at all. And BWT if className means the simple name and not FQCN then it's not very useful either.

I believe you are right, but I just checked my tests, I can't see any `@WithSpan` and as user I don't always read Quarkus implementation. I don't know whether for some user simple class and method name is enough, hence migration guide note.

mkouba commented:
> > Well, the point is that the behavior was undefined/undocumented. So it wasn't testing a contract but an implementation detail.
> 
> I had to rely on something when writing these tests. I don't say you established contract, but little note in migration guide for people like me would help. Thanks

We will add a note to the docs so that it's clear and testable.

> 
> > I mean, previously quarkus.scheduler.tracing.enabled=true implied that @WithSpan should be added to a scheduled method and the javadoc of @WithSpan.value() states: "If not specified, an appropriate default name should be created by auto-instrumentation. E.g. "className"."method"." Which is IMO not clear or deterministic at all. And BWT if className means the simple name and not FQCN then it's not very useful either.
> 
> I believe you are right, but I just checked my tests, I can't see any `@WithSpan` and as user I don't always read Quarkus implementation. I don't know whether for some user simple class and method name is enough, hence migration guide note.

`@WithSpan` was added automatically. So it's even more confusing ;-).

I've added a migration note here:
https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.7#scheduler---opentelemetry-tracing

michalvavrik commented:
Thanks!

brunobat commented:
Thanks @mskacelik. This is great stuff.
